File: day04/part1.py
_TESTCASE = """\
ecl:gry pid:860033327 eyr:2020 hcl:#fffffd
byr:1937 iyr:2017 cid:147 hgt:183cm

iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884
hcl:#cfa07d byr:1929

hcl:#ae17e1 iyr:2013
eyr:2024
ecl:brn pid:760753108 byr:1931
hgt:179cm

hcl:#cfa07d eyr:2025 pid:166559648
iyr:2011 ecl:brn hgt:59in
""".strip().split(
    "\n\n"
)


def compute(data):
    """
    >>> compute(_TESTCASE)
    2
    """
    required = frozenset(("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"))
    return sum(
        required
        <= {
            field.split(":")[0]
            for line in block.splitlines()
            for field in line.split(" ")
        }
        for block in data
    )


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")

    with input_path.open() as f:
        print(compute(f.read().strip().split("\n\n")))


if __name__ == "__main__":
    main()

File: day04/part1.hs
import Data.List
import System.Environment
import System.FilePath


-- implemented most of the stuff before i found out about haskell's "standard library"


type Block = [String]
type Entry = (String, String)
type Passport = [Entry]


required :: [String]
required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]


split :: Eq a => a -> [a] -> [[a]]
split d s = case span (/= d) s of
  (x, _ : ts) -> x : split d ts
  (x, []    ) -> x : []


-- https://stackoverflow.com/a/23979003
sameElems :: Eq a => [a] -> [a] -> Bool
sameElems xs ys = null (xs \\ ys) && null (ys \\ xs)


blocks :: [String] -> [Block]
blocks [] = []
blocks ls = a' : blocks b
 where
  (a, b) = case span (/= "") ls of
    ([], b) -> ([], tail b)
    (a , b) -> (a, b)
  a' = concatMap (split ' ') a


entry :: String -> Entry
entry s = case split ':' s of
  [a, b] -> (a, b)


passport :: Block -> Passport
passport xs = map entry xs


checkPassport :: Passport -> Bool
checkPassport p = intersect (map fst p) required `sameElems` required


compute :: [String] -> Int
compute ls = length . (filter checkPassport) . (map passport) . blocks $ ls


main :: IO ()
main = do
  exePath <- getExecutablePath
  content <- readFile (replaceFileName exePath "input.txt")
  putStrLn . show . compute . lines $ content
